feat: dependent feature toggles

daveleek opened the pull request:
# Description

Adds support for dependent features in .NET SDK. Also adds a warn once component

## Type of change

- [x] New feature (non-breaking change which adds functionality)


# How Has This Been Tested?

- [x] Running spec tests locally
- [x] Adding and running a suite of unit tests

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

kwasniew reviewed (commented):
Overall it looks very good. Just some questions:
* why do we have to write all the tests inside this repo. Isn't the specification covering everything that we need?
* can we add a test that shows that metrics are not called on the dependent features
* do we have impression events in .net SDK? If so, how do they behave on the dependent features?

daveleek commented:
Good questions @kwasniew!

> why do we have to write all the tests inside this repo. Isn't the specification covering everything that we need?

it's been convention so far to also add local unit tests, helps with development and debugging etc, I'll rope @sighphyre in for opinions though

> can we add a test that shows that metrics are not called on the dependent features

Coming right up!

> do we have impression events in .net SDK? If so, how do they behave on the dependent features?

sighphyre commented on src/Unleash/DefaultUnleash.cs:
```diff
@@ -155,9 +160,43 @@ public bool IsEnabled(string toggleName, UnleashContext context, bool defaultSet
+            var parentToggle = GetToggle(dependency.Feature);
+            if (parentToggle == null)
+            {
+                warnOnce.Warn(dependency.Feature, $"UNLEASH: Parent feature toggle {dependency.Feature} was not found in the cache, the evaluation of this dependency will always be false");
```
This will log once per parent that's disabled. So if you have toggle A and toggle B that both depend on toggle C, which is missing, once A is evaluated and this message shows,  the B will never get a warning message. Is that what we want? (Genuinely asking, I've heard about warn once but this is the first time I've seen it)

daveleek replied:
https://github.com/Unleash/unleash-client-node/pull/520

In the intro text:
> warning events for missing dependencies are reported once
> `this.warnedDependencies[missingDependency + name] = true;`

So no, should warn once per child and dependency it seems, I'll fix

daveleek replied:
I've addressed this now

sighphyre commented on src/Unleash/Internal/Dependency.cs:
```diff
@@ -0,0 +1,24 @@
+namespace Unleash.Internal {
+    public class Dependency {
```
Does this need to be public? Totally okay if it has to be or it makes sense to be but I'd like to hide stuff we don't want other people to permanently depend on - stupidly hard to undo public dependencies

daveleek replied:
It's a property of public class `FeatureToggle` so needs to be as accessible as it's containing object

sighphyre replied:
Okay, we chatted offline, it'd be great but this is much harder than simply this right now. I'm okay with this

sighphyre commented on src/Unleash/Utilities/WarnOnce.cs:
```diff
@@ -0,0 +1,29 @@
+        private readonly HashSet<string> seen = new HashSet<string>();
+
+        public WarnOnce(ILog logger)
+		{
```
Nit: Can we give this guy a little bit of autoformat love before we merge?

sighphyre commented on tests/Unleash.Tests/Integration/ClientSpecificationTests.cs:
```diff
@@ -38,7 +38,7 @@ static TestFactory()
             using (var client = new HttpClient())
             {
-                var csTestsVersion = "v4.3.3";
+                var csTestsVersion = "v4.5.1";
```
4.5.2 is out! Adds a test case, which is relevant to dependent features!

```suggestion
                var csTestsVersion = "v4.5.2";
```

sighphyre commented:
> Good questions @kwasniew!
> 
> > why do we have to write all the tests inside this repo. Isn't the specification covering everything that we need?
> 
> it's been convention so far to also add local unit tests, helps with development and debugging etc, I'll rope @sighphyre in for opinions though


I'd lean towards reducing the tests here. This is a really well covered feature in the client specs, I don't see much in your tests that aren't doing what the spec is doing. Usually I'm fussy about having tests that cover cases the spec doesn't for your particular implementation or in cases where a unit test is much easier to deal with than the spec. This is a pretty simple feature though

Things I'd like to see tested:

- What does metrics do here? (don't think this is covered, we could use some tests here)
- Does the response parse correctly and set the right defaults (this does look covered)
- Does the feature work as expected (covered through the client spec, I don't see anything you've done in the code that would require additional testing for the actual implementation)

daveleek commented:
@sighphyre 

> I'd lean towards reducing the tests here...  

> ... or in cases where a unit test is much easier to deal with than the spec. This is a pretty simple feature though

I guess I could trim some tests, but they've been valuable to verify understanding at development time and as debug tools so I'd like to keep a few in

> Things I'd like to see tested:
> What does metrics do here? (don't think this is covered, we could use some tests here)

Added a test for that

> What should impression events do (think you're working on this already)

Also added a test for this

> Does the response parse correctly and set the right defaults (this does look covered)

This should be covered, I have attempted to at least

>Does the feature work as expected (covered through the client spec, I don't see anything you've done in the code that would require additional testing for the actual implementation)

👍

sighphyre commented:
> @sighphyre
> 
> > I'd lean towards reducing the tests here...
> 
> > ... or in cases where a unit test is much easier to deal with than the spec. This is a pretty simple feature though
> 
> I guess I could trim some tests, but they've been valuable to verify understanding at development time and as debug tools so I'd like to keep a few in

If you find them useful I'm happy for them to stay. At some point we should get to the point where it's nice to use the client spec but that's definitely out of scope here (JS and Ruby are pretty good at this)

sighphyre reviewed (approved):
Happy if you are

kwasniew commented on tests/Unleash.Tests/Serialization/DynamicJsonSerializerTests.cs:
```diff
@@ -136,5 +136,19 @@ public void Serializes_ImpressionData_Property()
+            // Act
+            var deserialized = JsonConvert.DeserializeObject<ToggleCollection>(originalJson);
+            var toggle = deserialized.Features.First(f => f.Name == "enabled-child");
+            toggle.Should().NotBeNull();
```
missing //assert?

daveleek replied:
Good catch!

kwasniew commented on tests/Unleash.Tests/Internal/Dependent_Features_Tests.cs:
```diff
@@ -0,0 +1,610 @@
+            // Assert
+            unleash.services.MetricsBucket.StopCollectingMetrics(out var bucket);
+            var childMetrics = bucket.Toggles.Single(t => t.Key == "child-1").Value;
+            (childMetrics.No + childMetrics.Yes).Should().Be(1L);
```
personally I prefer explicit numbers in tests instead of calculations

daveleek replied:
Fair point, the explicit number was in the .Be part of the assertion - but I'll add a separate line for each to not sum the metrics no and yes!

kwasniew reviewed (approved):
I'm curious if you used client spec in development? You said that manually written tests were useful in development, but in Node SDK I used client spec for that. On the other hand I had to wade through hundreds lines of test code to find the metrics test I wanted to check. I'm trying to understand what exactly are we missing from the shared client spec tests. Are the .net tests easier to work with?

daveleek commented:
> I'm curious if you used client spec in development?

I did use that as well, but mostly after being done with large portions of code. And whenever something was wrong I could repro it with a specific unit test and debug from there.

> I'm trying to understand what exactly are we missing from the shared client spec tests. Are the .net tests easier to work with?

I'm not going to say the .NET tests are awesome, but for me it boils down a combination of two things:
- Habit - I'm used to unit tests showing me what I'm building and my understanding of it is sane)
- Convenience - Unit tests are (for now) easier to work with on a case by case basis. Can probably instrument the spec test helpers to give more granular control over tests to run, but I have no idea of the amount of work involved so short term easier to just keep doing what I'm doing

I'm also less comfortable around the implementation of the spec parsing and knowing what might be missing there but still providing false positives - so I use Unit tests as a way of reassuring myself

With that said I find the spec massively useful as an overview into where I'm at with the implementation 30/40/90% "done" / green board check off, and also as a source of knowledge/insight into the feature and how it's expected to work
